File: src/fragment/DimensionFragment.ts
import { Text } from "pixi.js";
import Fragment, { FragmentType, register } from "./Fragment";
import { StructEndecBuilder, PrimitiveEndecs } from 'KEndec';
import { Identifier } from "~/util";



export default class DimensionFragment extends Fragment {
    world: string;

    constructor(world: string) {
        super();

        this.world = world;
    }

    override toString(): string {
        return this.world.replace(/^[^a-zA-Z0-9]+/, "");
    }

    override type(): FragmentType<DimensionFragment> {
        return DIMENSION;
    }
}

const DIMENSION = register("dimension", 0xdd55bb, 
    StructEndecBuilder.of1(
        Identifier.ENDEC
            .xmap((ident) => ident.toString(), Identifier.of)
            .fieldOf("block", (fragment: DimensionFragment) => fragment.world),
        (str) => new DimensionFragment(str) 
    )
);

File: src/jsMain/kotlin/me/maplesyrum/tricksterstudio/spell/fragment/DimensionFragment.kt
package me.maplesyrum.tricksterstudio.spell.fragment

import me.maplesyrum.tricksterstudio.Identifier
import tree.maple.kendec.StructEndec
import tree.maple.kendec.impl.StructEndecBuilder


class DimensionFragment(val world: String) : Fragment() {

    override fun type(): FragmentType<*> {
        return FragmentType.DIMENSION
    }

    override fun toString(): String {
        return world.replace(Regex("^[^a-zA-Z0-9]+"), "")
    }

    companion object {
        var ENDEC: StructEndec<DimensionFragment> = StructEndecBuilder.of(
            Identifier.ENDEC
                .xmap({ id -> id.toString() }, { str -> Identifier.of(str) })
                .fieldOf("block", DimensionFragment::world)
        ) { str -> DimensionFragment(str) }
    }
}
